### ISSUE TITLE
Import rubrics between assessment questions

### ISSUE THREAD

**Issue by jonatanschroeder** [PrairieLearn/PrairieLearn#9616]
Currently rubrics must be populated manually in the UI, there is no way to copy them from one question to another. #7563 has a possible implementation of this feature, but it would likely make more sense to provide the means to copy directly the rubrics of one question to another (subject to the discussion in #8901).

**Comment by reteps** [comment]
This has been requested a couple times on Slack.

**Comment by Victorsss-Orz** [comment]
The UI might get messy if we want to allow quick importing from one assessment question to any assessment question because there are too many. Maybe we can implement something like "import from previous assessment questions that are the same question as this"?

**Comment by jonatanschroeder** [comment]
One option: the "Import rubric" button could be a dropdown that selects "From a file" or "From another assessment". If the latter is selected, a modal could be shown with all assessments (other than the current one) that have the same question and have a rubric. This list could be shown grouped by course instance in descending order of course instance date (for some interpretation of "date"), so the most recent option is preferred.

**Comment by ffund** [comment]
I realize this may be at odd with how rubrics are implemented. But I feel like rubrics "belong" to a particular question, and they should "live" inside the question directory as files. I want to do things like: 

- version control my rubrics in GitHub, track changes, revert if a grader does something dumb
- share a course repo with someone who's interested in adopting some of the material, and share the rubrics alongside the questions they belong to
- tell my preferred LLM assistant to "modify all the rubrics in these directories to change language X to Y"
- duplicate a question and its rubric is also in the copy
- edit my rubrics in my preferred code editor (no offense to the rubric editing UX, it's actually pretty nice, I'm just much more efficient in code editor.)

If the same rubrics used for manual grading are also used for a future AI grading feature, then all of these items become even more important.

I can see the concern of potentially needing to maintain multiple rubrics for the same question, maybe there could be a "rubrics" subdirectory inside the question directory, with multiple JSON rubrics?

To me, a major selling point of PL is that basically everything I create (that's not student data) lives in a repo. I don't know why rubrics should be an exception.

**Comment by firasm** [comment]
I agree with @ffund almost completely but in my mind, there's (at least) one implementation detail that makes it challenging for rubrics to be files that belong with the question. 

And that's the ability for rubrics to be edited on the fly and synchronized between different graders. If rubrics were files, I can't imagine how to have multiple graders edit it without having a commit/push operation (and the problem of how to flag to other graders that there's a rubric update without constantly fetching).

That said - I wish rubrics were files too that could be edited live but I think git can't be coerced to work that way.

**Comment by nwalters512** [comment]
To rattle off some problems here, and reasons why we didn't build things this way:

- Course sync operations are slooooow at the moment. This is related to what Firas said. Putting Git, GitHub, and the syncing process in the path of rubric modifications would make for a worse UX.
- Rubric items would need persistent identifiers that are independent of the text content of them. Maybe not a huge hurdle in practice, but it means if you're sitting there editing a rubric, you'd need to add UUIDs or something like that to every new item.
- You'd need to define what does actually happen when you update a rubric:
  - If you used the question last term, do you retroactively update all scores? How about if you used it last week on a homework?
  - How do you implement the "Require all graded submissions to be manually graded/reviewed" checkbox? You can't prompt for input like that during a sync.
- What if you do actually want rubrics to differ from term to term, or between assessments in the same term?
- How does this work with shared questions, where the consuming course doesn't have write access to the original one?

Now, the above is all assuming that the desire is to treat Git as the source of truth at all times. There could be a middle ground where you can (manually, from the UI) "export" a rubric to a question directory, and "import" it back into assessment questions. This doesn't solve the "my TA did something dumb and I want to undo it" problem, but it does solve most (all?) of the others.

**Comment by ffund** [comment]
I would be happy with that middle ground. 

* It still treats the rubric as a first-class OER resource that I own and can work with outside of PL, not just something that makes grading more convenient. 
* It also helps with having to maintain multiple rubrics for the same question (e.g. different grading when you assign the same question to an undergrad class or grad class) - when you import, you can choose which rubric to import from the "rubrics" directory.

If implemented this way, it would be very useful to have some visual indicator telling me whether the current rubric is synced/not synced with the file.
